### ISSUE TITLE
Large numbers are represented in scientific notation

### ISSUE THREAD

**Issue by cemremengu** [oracle/node-oracledb#842]
We have large numbers as `ID` columns in our database. When they are selected, the output in in scientific notation like ` 6.000370005658599e+21`. I need the whole value (it can be long or string). 

I know that I can use `fetchAsString` for `NUMBER` type but I don't want _all_ numbers as strings.

Is there a way to get around this?

**Comment by anthony-tuininga** [comment]
You can use [fetchInfo](https://github.com/oracle/node-oracledb/blob/master/doc/api.md#propexecfetchinfo) in order to specify for particular columns instead of all numeric values.

**Comment by cemremengu** [comment]
Thanks, this works but not really practical since I have to specify any rows like that up-front and I am working in a dynamic context. Maybe if it supported pattern matching, it could have been more useful.

Any other solutions besides this?

**Comment by anthony-tuininga** [comment]
Not currently. Perhaps @dmcghan may have something up his sleeves, though. :-) The only other possibility I can consider is a bit of a hack -- adjust the query to add "1 = 0" so that no rows are actually returned and examine the metadata, then make the query again with the appropriate fetchInfo directive.

There has been some discussion of providing a hook at the point where the metadata is known but before the fetch areas have been established -- which would give you the opportunity to adjust these things without the hack noted above. But that is as far it has progressed so far, unfortunately!

**Comment by cemremengu** [comment]
Thank you for your help! May I ask if you know why the return value is in scientific notation?

**Comment by anthony-tuininga** [comment]
Sure. Node.js uses double floating point numbers as its native number type. As such it can only represent numbers that require 53 or less binary digits -- in other words 2 ^ 53 (9007199254740992). Numbers larger than this will be truncated in some fashion. You can see this by attempting to set a Node.js variable to a larger value.

```javascript
var x = 9007199254740993;
console.log(x);
x = 9007199254740993123456;
console.log(x);
```

You will get this output:

9007199254740992
9.007199254740993e+21

Note that the first result is the wrong value. The second one is too, but it shows it in scientific notation so its harder to see that fact.

So, if you want to use numbers larger than 9007199254740992 you will *have* to use a string to get them from the database and then you will have to use a JavaScript library like bigint to actually manipulate them (if you need to manipulate them as numbers).

**Comment by cemremengu** [comment]
Now that's a real bummer... Thanks for the info and your work on the this library!

**Comment by cjbj** [comment]
A few random comments:

- you may currently hit https://github.com/oracle/node-oracledb/issues/833 with numbers unless you are fetching as strings.  A fix is coming soon.
- to get metadata you don't need to add "1=0": you can use a resultset and close it immediately (after looking at the metadata) without fetching any rows.
- I will update [doc](https://github.com/oracle/node-oracledb/blob/master/doc/api.md#numberhandling) to mention the range (this was also noted by @anthony-tuininga when the above bugfix was reviewed)

**Comment by cemremengu** [comment]
@cjbj Hoping that the fix in #833 solve the problem I am having as well? I have been having nightmares since yesterday 😄

**Comment by cjbj** [comment]
@cemremengu even with the fix for #833, you may still want to use fetchAsString or do all calculations in PL/SQL & SQL to avoid any range problems or conversion quirks.

**Comment by anthony-tuininga** [comment]
> @cjbj Hoping that the fix in #833 solve the problem I am having as well? I have been having nightmares since yesterday.

I'm sorry to have to tell you that it definitely **won't** fix your issues. You're attempting to use numbers that exceed 2 ^ 53 and those can't be represented properly in Node.js numbers. :-(
